### DISCUSSION TITLE
:matches-css() no longer works in v1.45.0

### DISCUSSION THREAD

**Discussion by esperecyan** [uBlockOrigin/uBlock-issues#2354] [Q&A] [answered]
With Firefox 106.0.5 + uBlock Origin 1.45.0, :matches-css() no longer works, has the spec changed?
Is there any way to get the following filters to work with 1.45.0?

### Filter

```
twitter.com##[data-testid="tweet"] [aria-labelledby] div:style(filter: none !important)
twitter.com##[data-testid="tweet"] [aria-labelledby] :matches-css(background-color: rgba(0, 0, 0, 0.5)) svg:has(~ div span:has-text(The Tweet author flagged this Tweet as showing sensitive content.))
twitter.com##[data-testid="tweet"] [aria-labelledby] :matches-css(background-color: rgba(0, 0, 0, 0.5)):style(top: unset !important; left: unset !important; padding: unset !important)
twitter.com##[data-testid="tweet"] [aria-labelledby] :matches-css(background-color: rgba(0, 0, 0, 0.5)) div:style(margin: unset !important)
twitter.com##[data-testid="tweet"] [aria-labelledby] :matches-css(background-color: rgba(0, 0, 0, 0.5)) span:has-text(The Tweet author flagged this Tweet as showing sensitive content.)
twitter.com##[data-testid="tweet"] [aria-labelledby] :matches-css(background-color: rgba(0, 0, 0, 0.5)) div:not(:only-of-type):has(span:has-text(Show))
```

### Expected Result (v1.44.4)
![](https://user-images.githubusercontent.com/3625160/201019686-9bd2e6ad-a8d1-47b4-b5f2-192547440005.png)

### Actual Result (v1.45.0)
![](https://user-images.githubusercontent.com/3625160/201019703-c8cbcf20-a0bb-4cb1-8eae-c73be833cb50.png)

### Reproduction Procedure
1. Installs uBlock Origin 1.45.0.  
   https://addons.mozilla.org/firefox/addon/ublock-origin/
1. Adds the above filter to My Filters in uBlock Origin Settings.
1. Open https://twitter.com/esperecyan/status/1590592413885952000 .
1. You will see the same "Actual Result" as above.
1. Downgrades to uBlock Origin 1.44.4.  
   https://addons.mozilla.org/ja/firefox/addon/ublock-origin/versions/
1.  You will see the same "Expected Result" as above

**Comment by ghost** [comment]
Do  you have any different example? I don't see the warning when logged-in, and post is not displayed at all when not logged-in.

**Reply by esperecyan** [reply]
I have written the simple sample.

````
jshell.net##body :matches-css(background-color: rgba(100, 100, 100, 0.5)):style(color: red)
````

https://jsfiddle.net/esperecyan/48hs675c/

```html
<p>test</p>
```

```css
p {
	background-color: rgba(100, 100, 100, 0.5);
}
```

In v1.14.4, the word “test” in the lower right frame is red, but in v1.15.0 it remains black.

**Reply by krystian3w** [reply]
Now needed are:

 `body *:matches-css(`

 `[aria-labelledby] *:matches-css(`

So bug is in parser based on CSStree?

**Reply by esperecyan** [reply]
Thank you!　The filter worked as expected!

```
twitter.com##[data-testid="tweet"] [aria-labelledby] div:style(filter: none !important)
twitter.com##[data-testid="tweet"] [aria-labelledby] *:matches-css(background-color: rgba(0, 0, 0, 0.5)) svg:has(~ div span:has-text(The Tweet author flagged this Tweet as showing sensitive content.))
twitter.com##[data-testid="tweet"] [aria-labelledby] *:matches-css(background-color: rgba(0, 0, 0, 0.5)):style(top: unset !important; left: unset !important; padding: unset !important)
twitter.com##[data-testid="tweet"] [aria-labelledby] *:matches-css(background-color: rgba(0, 0, 0, 0.5)) div:style(margin: unset !important)
twitter.com##[data-testid="tweet"] [aria-labelledby] *:matches-css(background-color: rgba(0, 0, 0, 0.5)) span:has-text(The Tweet author flagged this Tweet as showing sensitive content.)
twitter.com##[data-testid="tweet"] [aria-labelledby] *:matches-css(background-color: rgba(0, 0, 0, 0.5)) div:not(:only-of-type):has(span:has-text(Show))
```

**Reply by krystian3w** [reply]
@gorhill

**Reply by Yuki2718** [reply]
Off-topic but each `##[data-testid="tweet"] [aria-labelledby] *` has around 200 nodes to evaluate and you have 5 such filters:
<details>
<summary>Screenshot</summary>

![twitter](https://user-images.githubusercontent.com/58900598/201078282-12f2046f-d222-4ff3-82a6-a122ff42b42d.png)

</details>
This is a bit concerning performance-wise. You may be able to narrow down the scope of them to avoid things like https://www.reddit.com/r/uBlockOrigin/comments/kg7224/youtube_homepage_loading_high_cpu/ (the less nodes, the better in general).

**Reply by krystian3w** [reply]
I think older parsers also check 200+ nodes so his CPU should alive.

**Reply by esperecyan** [reply]
Twitter's HTML is encrypted, and I think I have no choice but to write it this way in order to make it less susceptible to changes in Twitter's specifications.

**Reply by Yuki2718** [reply]
Fine in that case.

**Reply by ghost** [reply]
> has around 200 nodes to evaluate

Isn't this dependent on thread size? uBO can eventually stop blocking these when time limit is exhausted.

**Reply by krystian3w** [reply, marked as answer]
https://github.com/gorhill/uBlock/commit/4d4a5f380799b64c167e42331a37d7be2f917214

**Comment by krystian3w** [comment]
https://github.com/gorhill/uBlock/commit/4d4a5f380799b64c167e42331a37d7be2f917214

#2355
